runtime/src/main/java/org/corfudb/protocols/wireprotocol/NodeState.java

```diff
@@ -1,81 +1,101 @@
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeConnectivity.NodeConnectivityType;
 import org.corfudb.runtime.view.Layout;
-import org.corfudb.util.NodeLocator;
 
 /**
  * Contains a Node's state:
  * Sequencer state - ready/not_ready.
  * connectivity status - Node's connectivity with every other node in the layout.
+ * <p>
+ * For instance, node a fully connected to all nodes:
+ * {"a": {"endpoint": "a", "connectivity":{"a": true, "b": true, "c": true}}}
+ * <p>
  * Created by zlokhandwala on 11/2/18.
  */
 @Data
 @Builder
 @AllArgsConstructor
+@EqualsAndHashCode
 public class NodeState implements ICorfuPayload<NodeState> {
 
     public static final long INVALID_HEARTBEAT_COUNTER = -1L;
 
-    /**
-     * Current node's Endpoint.
-     */
-    private final NodeLocator endpoint;
+    private final NodeConnectivity connectivity;
 
-    private final long epoch;
-
-    /**
-     * Heartbeat counter.
-     */
-    private final long heartbeatCounter;
+    private final HeartbeatTimestamp heartbeat;
 
     /**
      * Sequencer metrics of the node.
      */
     private final SequencerMetrics sequencerMetrics;
 
-    /**
-     * Node's view of the cluster.
-     */
-    private final Map<String, Boolean> connectivityStatus;
-
     public NodeState(ByteBuf buf) {
-        endpoint = NodeLocator.parseString(ICorfuPayload.fromBuffer(buf, String.class));
-        epoch = ICorfuPayload.fromBuffer(buf, Long.class);
-        heartbeatCounter = ICorfuPayload.fromBuffer(buf, Long.class);
+        connectivity = ICorfuPayload.fromBuffer(buf, NodeConnectivity.class);
+        heartbeat = ICorfuPayload.fromBuffer(buf, HeartbeatTimestamp.class);
         sequencerMetrics = ICorfuPayload.fromBuffer(buf, SequencerMetrics.class);
-        connectivityStatus = ICorfuPayload.mapFromBuffer(buf, String.class, Boolean.class);
     }
 
     @Override
     public void doSerialize(ByteBuf buf) {
-        ICorfuPayload.serialize(buf, endpoint.toString());
-        ICorfuPayload.serialize(buf, epoch);
-        ICorfuPayload.serialize(buf, heartbeatCounter);
+        ICorfuPayload.serialize(buf, connectivity);
+        ICorfuPayload.serialize(buf, heartbeat);
         ICorfuPayload.serialize(buf, sequencerMetrics);
-        ICorfuPayload.serialize(buf, connectivityStatus);
+    }
+
+    public static NodeState getUnavailableNodeState(String endpoint){
+        return new NodeState(
+                NodeConnectivity.unavailable(endpoint),
+                new HeartbeatTimestamp(Layout.INVALID_EPOCH, INVALID_HEARTBEAT_COUNTER),
+                SequencerMetrics.UNKNOWN
+        );
+    }
+
+    public static NodeState getNotReadyNodeState(String endpoint, long epoch, long counter){
+        return new NodeState(
+                NodeConnectivity.notReady(endpoint),
+                //Layout.INVALID_EPOCH, INVALID_HEARTBEAT_COUNTER - default values for heartbeat
+                new HeartbeatTimestamp(epoch, counter),
+                SequencerMetrics.UNKNOWN
```

Comment:
Shouldn't this be NOT_READY?

Reply:
We actually don't know, it could be ready or not, we just don't know